File: Chapter01/Activity1_02/Activity1_02.scala
package Activity1_02

import org.apache.spark.rdd.RDD
import org.apache.spark.sql.SparkSession
import Utilities01.HelperScala.{createSession, novellaLocation}

object Activity1_02 {

  def main(args: Array[String]): Unit = {

    val session: SparkSession = createSession(2, "Zipf Validation")
    val lines: RDD[String] = session.sparkContext.textFile(novellaLocation)

    val tokens: RDD[String] = lines
      .flatMap(_.trim().toLowerCase.split("\\W+"))
      .filter(_.nonEmpty)

    val countsPerToken: RDD[(String, Int)] = tokens
      .map(token => (token, 1))
      .reduceByKey((count1, count2) => count1 + count2, 2)

    val sortedCountsToken = countsPerToken
      .map(_.swap) // swap is equivalent to `pair =>(pair._2, pair._1)`
      .sortByKey(ascending = false)

    sortedCountsToken.saveAsTextFile(s"zipfsorted")
  }

}

File: Chapter01/Activity1_02/Activity1_02.py
from pyspark.rdd import RDD
import re
from operator import add
from utilities01_py.helper_python import *

if __name__ == "__main__":
    session: SparkSession = create_session(2, "Zipf validation")
    lines: RDD = session.sparkContext.textFile(novella_location)

    tokens = lines \
        .flatMap(lambda line: re.split('\\W+', line.strip().lower())) \
        .filter(lambda record: record != '')

    counts_per_token = tokens \
        .map(lambda token: (token, 1)) \
        .reduceByKey(add)

    sorted_counts = counts_per_token \
        .map(lambda tuple: (tuple[1], tuple[0])) \
        .sortByKey(ascending=False)

    sorted_counts.saveAsTextFile('zipfsorted_py')
